plugins/badges-backend/src/service/router.ts

```diff
@@ -125,25 +187,255 @@ export async function createRouter(
+        const hash = crypto
+          .createHash('sha256')
+          .update(`${kind}:${namespace}:${name}:${salt}`)
+          .digest('hex');
+
+        return res.status(200).json({ hash });
+      },
+    );
+
+    router.use(errorHandler());
+
+    return router;
+
+    // If the obfuscation is disabled, use the previously implemented routes
+    // eslint-disable-next-line no-else-return
+  } else {
+    router.get(
+      '/entity/:namespace/:kind/:name/badge-specs',
+      async (req, res) => {
+        const token = await tokenManager.getToken();
+        const { namespace, kind, name } = req.params;
+        const entity = await catalog.getEntityByRef(
+          { namespace, kind, name },
+          token,
+        );
+        if (!entity) {
+          throw new NotFoundError(
+            `No ${kind} entity in ${namespace} named "${name}"`,
+          );
+        }
+
+        const specs = [];
+        for (const badgeInfo of await badgeBuilder.getBadges()) {
+          const context: BadgeContext = {
+            badgeUrl: await getBadgeUrl(namespace, kind, name, badgeInfo.id),
+            config: options.config,
+            entity,
+          };
+
+          const badge = await badgeBuilder.createBadgeJson({
+            badgeInfo,
+            context,
+          });
+          specs.push(badge);
+        }
+
+        res.status(200).json(specs);
+      },
+    );
+
+    router.get(
+      '/entity/:namespace/:kind/:name/badge/:badgeId',
+      async (req, res) => {
+        const { namespace, kind, name, badgeId } = req.params;
+        const token = await tokenManager.getToken();
+        const entity = await catalog.getEntityByRef(
+          { namespace, kind, name },
+          token,
+        );
+        if (!entity) {
+          throw new NotFoundError(
+            `No ${kind} entity in ${namespace} named "${name}"`,
+          );
+        }
+
+        let format =
+          req.accepts(['image/svg+xml', 'application/json']) || 'image/svg+xml';
+        if (req.query.format === 'json') {
+          format = 'application/json';
+        }
+
+        const badgeOptions = {
+          badgeInfo: { id: badgeId },
+          context: {
+            badgeUrl: await getBadgeUrl(namespace, kind, name, badgeId),
+            config: options.config,
+            entity,
+          },
+        };
+
+        let data: string;
+        if (format === 'application/json') {
+          data = JSON.stringify(
+            await badgeBuilder.createBadgeJson(badgeOptions),
+            null,
+            2,
+          );
+        } else {
+          data = await badgeBuilder.createBadgeSvg(badgeOptions);
+        }
+
+        res.setHeader('Content-Type', format);
+        res.status(200).send(data);
+      },
+    );
+
+    router.use(errorHandler());
+
+    return router;
+  }
+
+  // This function generate a table that maps the hash to the namespace/kind/name triplet
+  async function generateHashLookupTable() {
```

Comment:
Sorry I realise that we need to compute a hash for each entity with the current implementation so my first comment does not apply.

Could we instead use the Entity `uid`(uuid) as the "hash". That would get us to one request to the catalog and no need to keep internal state. That would be hard to guess/enumerate :)

Reply:
Yes I thought about some DB persistence, but I was thinking in a Gitops way, like any disaster in the instance / DB corruption / re installation... Would need any badges to be updated because of the moving uuid. While the hash mechanism is tied to the entity metadata, stored in the catalog-info file.

Reply:
Yes I thought about some DB persistence, but I was thinking in a Gitops way, like any disaster in the instance / DB corruption / re installation... Would need any badges to be updated in every external source because of the moving parts like uuid. While the hash mechanism is tied to the entity metadata, stored in the catalog-info file. (As actually in fact)

Although [in the documentation](https://backstage.io/docs/features/software-catalog/descriptor-format/#uid-output), it's clearly mentioned

> Note that uid values are not to be seen as stable, and should not be used as external references to an entity

Reply:
It's true that the uid is not stable if you unregister/register or delete the database table.

I discussed the current implementation with @Rugvip, and we suggest to instad have a database table for the the obfuscated badges. Storing the entity triplet and a generated UUID mapping will remove the need to query all entities periodically and lead to a snappier experience.